src/main/scala/beam/agentsim/agents/ridehail/RideHailUtils.scala

```diff
@@ -18,51 +19,48 @@ object RideHailUtils {
     transportNetwork: TransportNetwork
   ): BeamLeg = {
 
-    if (stopTime < originalBeamLeg.startTime || stopTime >= originalBeamLeg.endTime)
-      return originalBeamLeg
+    if (stopTime < originalBeamLeg.startTime || stopTime >= originalBeamLeg.endTime) {
+      originalBeamLeg
+    } else {
+
+      val duration = stopTime - originalBeamLeg.startTime
+
+      val pctTravelled: Double = duration / originalBeamLeg.duration
 
-    val pctTravelled = (stopTime - originalBeamLeg.startTime) / originalBeamLeg.duration
-    val distanceOfNewPath = originalBeamLeg.travelPath.distanceInM * pctTravelled
+      val distanceOfNewPath = originalBeamLeg.travelPath.distanceInM * pctTravelled
 
-    var updatedLinkIds: Vector[Int] = Vector()
+      var endPointLocation: Coord = originalBeamLeg.travelPath.endPoint.loc
 
-    var endPointLocation = originalBeamLeg.travelPath.endPoint.loc
-    var updatedDistanceInMeters = distanceOfNewPath
+      var updatedDistanceInMeters: Double = distanceOfNewPath
 
-    var linkIds = updatedLinkIds
-    for (linkId <- originalBeamLeg.travelPath.linkIds) {
-      linkIds = linkIds :+ linkId
-      val distance = getDistance(linkIds, transportNetwork)
+      val linkIds = ListBuffer[Int]()
 
-      breakable {
-        if (distanceOfNewPath < distance) {
-          break
-        } else {
-          endPointLocation = GeoUtils.getR5EdgeCoord(linkId, transportNetwork)
-          updatedDistanceInMeters = distance
-          updatedLinkIds = linkIds
+      var distance: Double = 0D
+
+      for (linkId: Int <- originalBeamLeg.travelPath.linkIds) {
+        linkIds += linkId
+        distance += transportNetwork.streetLayer.edgeStore.getCursor(linkId).getLengthM
+
+        breakable {
+          if (distanceOfNewPath < distance) {
+            break
+          } else {
+            endPointLocation = GeoUtils.getR5EdgeCoord(linkId, transportNetwork)
+            updatedDistanceInMeters = distance
+          }
         }
       }
-    }
 
-    val updatedEndPoint = SpaceTime(endPointLocation, stopTime.toLong)
-    val updatedTravelPath = originalBeamLeg.travelPath.copy(
-      linkIds = updatedLinkIds,
-      endPoint = updatedEndPoint,
-      distanceInM = updatedDistanceInMeters
-    )
-    val updatedDuration = (stopTime - originalBeamLeg.startTime).toLong
+      val updatedEndPoint = SpaceTime(endPointLocation, stopTime.toLong)
 
-    originalBeamLeg.copy(duration = updatedDuration, travelPath = updatedTravelPath)
-  }
+      val updatedTravelPath = originalBeamLeg.travelPath.copy(
+        linkIds = linkIds.toVector,
```

Comment:
`linkIds.toVector` will allocate new buffer for `Vector`. Why just don't change `BeamPath. linkIds` type to `Seq[Int]` and do not need any conversion between collections . Also I would go with `ArrayBuffer` instead of `ListBuffer` due to data locality (more data will be in cache)

Reply:
Hi @REASY ,
According to the documentation `https://www.scala-lang.org/api/current/scala/collection/mutable/ListBuffer.html`
```Scala
def toList: immutable.List[A]
```
> Converts this buffer to a list. Takes constant time. The buffer is copied lazily, the first time it is mutated.
Constant time means doesn´t allocate new buffer

Reply:
For some reason, I have `toList` in mind :-)